<a href="https://colab.research.google.com/github/dspuliaiev/Data_Science/blob/master/Hw11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from keras.datasets import imdb
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Embedding, SimpleRNN, LSTM, Bidirectional, Dense, Dropout

# Завантаження даних
max_features = 20000  # Кількість слів
max_len = 200  # Максимальна довжина рецензії

(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=max_features)

# Заповнення або обрізання рецензій до однакової довжини
X_train = sequence.pad_sequences(X_train, maxlen=max_len)
X_test = sequence.pad_sequences(X_test, maxlen=max_len)


In [23]:
## RNN

model_rnn = Sequential()
model_rnn.add(Embedding(max_features, 128, input_length=max_len))
model_rnn.add(SimpleRNN(128, return_sequences=False))
model_rnn.add(Dense(1, activation='sigmoid'))

model_rnn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history_rnn = model_rnn.fit(X_train, y_train, epochs=3, batch_size=128, validation_split=0.2)

rnn_acc = model_rnn.evaluate(X_test, y_test)[1]


Epoch 1/3
157/157 [==============================] - 44s 269ms/step - loss: 0.6513 - accuracy: 0.6007 - val_loss: 0.5090 - val_accuracy: 0.7572
Epoch 2/3
157/157 [==============================] - 36s 230ms/step - loss: 0.4190 - accuracy: 0.8118 - val_loss: 0.4867 - val_accuracy: 0.7902
Epoch 3/3
782/782 [==============================] - 14s 18ms/step - loss: 0.5635 - accuracy: 0.7652


In [24]:
## LSTM

model_lstm = Sequential()
model_lstm.add(Embedding(max_features, 128, input_length=max_len))
model_lstm.add(LSTM(128, return_sequences=False))
model_lstm.add(Dense(1, activation='sigmoid'))

model_lstm.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history_lstm = model_lstm.fit(X_train, y_train, epochs=3, batch_size=128, validation_split=0.2)

lstm_acc = model_lstm.evaluate(X_test, y_test)[1]


Epoch 1/3
157/157 [==============================] - 19s 104ms/step - loss: 0.4836 - accuracy: 0.7582 - val_loss: 0.3237 - val_accuracy: 0.8666
Epoch 2/3
157/157 [==============================] - 12s 77ms/step - loss: 0.2348 - accuracy: 0.9108 - val_loss: 0.3454 - val_accuracy: 0.8536
Epoch 3/3
782/782 [==============================] - 4s 5ms/step - loss: 0.3880 - accuracy: 0.8560


In [25]:
## Двостороння LSTM

model_bilstm = Sequential()
model_bilstm.add(Embedding(max_features, 128, input_length=max_len))
model_bilstm.add(Bidirectional(LSTM(128, return_sequences=False)))
model_bilstm.add(Dense(1, activation='sigmoid'))

model_bilstm.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history_bilstm = model_bilstm.fit(X_train, y_train, epochs=3, batch_size=128, validation_split=0.2)

bilstm_acc = model_bilstm.evaluate(X_test, y_test)[1]


Epoch 1/3
157/157 [==============================] - 22s 117ms/step - loss: 0.4254 - accuracy: 0.7969 - val_loss: 0.4234 - val_accuracy: 0.8394
Epoch 2/3
157/157 [==============================] - 14s 90ms/step - loss: 0.2131 - accuracy: 0.9205 - val_loss: 0.3184 - val_accuracy: 0.8726
Epoch 3/3
782/782 [==============================] - 7s 8ms/step - loss: 0.3556 - accuracy: 0.8624


In [26]:
## Глибока LSTM

model_deep = Sequential()
model_deep.add(Embedding(max_features, 128, input_length=max_len))
model_deep.add(LSTM(128, return_sequences=True))
model_deep.add(Dropout(0.5))
model_deep.add(LSTM(128, return_sequences=False))
model_deep.add(Dense(1, activation='sigmoid'))

model_deep.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history_deep = model_deep.fit(X_train, y_train, epochs=3, batch_size=128, validation_split=0.2)

deep_acc = model_deep.evaluate(X_test, y_test)[1]


Epoch 1/3
157/157 [==============================] - 21s 113ms/step - loss: 0.4361 - accuracy: 0.7893 - val_loss: 0.2993 - val_accuracy: 0.8758
Epoch 2/3
157/157 [==============================] - 14s 86ms/step - loss: 0.2046 - accuracy: 0.9231 - val_loss: 0.3978 - val_accuracy: 0.8406
Epoch 3/3
782/782 [==============================] - 6s 8ms/step - loss: 0.4088 - accuracy: 0.8456


In [27]:
## Порівняння результатів

print(f"Simple RNN Accuracy: {100*rnn_acc:.1f} %")
print(f"LSTM Accuracy: {100*lstm_acc:.1f} %")
print(f"Bidirectional LSTM Accuracy: {100*bilstm_acc:.1f} %")
print(f"Deep LSTM Accuracy: {100*deep_acc:.1f} %")

Simple RNN Accuracy: 76.5 %
LSTM Accuracy: 85.6 %
Bidirectional LSTM Accuracy: 86.2 %
Deep LSTM Accuracy: 84.6 %


## Висновки:

Простий RNN: Точність на тестових даних: 76.52%
Проста RNN показує прийнятну точність, але значно відстає від більш складних моделей.


LSTM: Точність на тестових даних: 86.24%
Модель LSTM значно покращує продуктивність у порівнянні з простою RNN.


Двостороння LSTM: Точність на тестових даних: 85.64%
Двостороння LSTM враховує контекст як з минулого, так і з майбутнього, що дає високий рівень точності.


Глибока LSTM: Точність на тестових даних: 84.56%
Глибока LSTM з двома шарами показує хорошу точність, оскільки здатна захоплювати більш складні патерни у даних.


При збільшенні кількості епох, було помічено, що на тренувальних даних точність продовжує зростати, але на валідаційних даних починає знижуватись. Це свідчить про перенавчання моделі, коли вона занадто добре підлаштовується під тренувальні дані, але втрачає здатність узагальнювати на нові дані. Зростання валідаційних втрат є індикатором цього явища.